Data/ByteString/Streaming/Char8.hs

```diff
@@ -710,20 +711,215 @@ putStrLn :: MonadIO m => ByteString m r -> m r
+                    !byte <- peek p
+                    let !w = byte - 0x30
+                        !d = fromIntegral w
+                    if | w > 9
+                         -- No more digits
+                         -> return (b, a, True)
+                       | a < maxq
+                         -- Look for more
+                         -> go (p `plusPtr` 1) (b + 1) (a * 10 + d)
+                       | a > maxq
+                         -- overflow
+                         -> return (b, a, False)
+                       | d <= maxr
+                         -- Ideally this will be the last digit
+                         -> go (p `plusPtr` 1) (b + 1) (a * 10 + d)
+                       | otherwise
+                         -- overflow
+                         -> return (b, a, False)
+
+        -- | Plausible success, sanity-check!
+        result !nbytes !acc str
+            | nbytes > 0, !i <- w2int acc = return $! Compose $ Just i :> str
+            | otherwise = overflow nbytes acc str
+
+        -- This assumes that @negate . fromIntegral@ correctly produces
+        -- @minBound :: Int@ when given its positive 'Word' value as an
+        -- input.  This is true in both 2s-complement and 1s-complement
+        -- arithmetic, so seems like a safe bet.  Tests cover this case,
+        -- though the CI may not run on sufficiently exotic CPUs.
+        w2int !n | positive = fromIntegral n
+                 | otherwise = negate $! fromIntegral n
+
+        -- | Reconstruct any consumed input, and report failure
+        overflow 0 _ str = case signByte of
+            Nothing -> return $ Compose $ Nothing :> str
+            Just w  -> return $ Compose $ Nothing :> Chunk (B.singleton w) str
+        overflow !nbytes !acc str =
+            let !c = overflowBytes nbytes acc
+             in return $! Compose $ Nothing :> Chunk c str
+
+        -- | Reconstruct an @nbytes@-byte prefix consisting of digits
+        -- from the accumulated value @acc@, with sufficiently many
+        -- leading zeros to match the original input length.  This
+        -- relies on decimal numbers (leading zeros aside) having a
+        -- unique representation.  Doing this for potentially mixed-case
+        -- hexadecimal input would require holding on to the input data,
+        -- which would noticeably hurt performance.
+        overflowBytes :: Int -> Word -> B.ByteString
+        overflowBytes !nbytes !acc =
+            B.unsafeCreate (nbytes + signlen) $ \p -> do
+                let end = p `plusPtr` (signlen - 1)
+                    ptr = p `plusPtr` (nbytes + signlen - 1)
+                go end ptr acc
+                mapM_ (poke p) signByte
+          where
+            signlen = if signByte == Nothing then 0 else 1
+
+            go :: Ptr Word8 -> Ptr Word8 -> Word -> IO ()
+            go end !ptr !_ | end == ptr = return ()
+            go end !ptr !a = do
+                let (q, r) = a `quotRem` 10
```

Comment:
GHC does not optimize division by a constant. I suggest to replace division by multiplication manually:

```haskell
-- | In 64-bit case
-- q = floor(floor((2^67+2)/10 * n) / 2^67)
--   = floor((2^67+2)/10 * n/2^67)
--   = floor(n/10 + (n/2^64 * 1/4)/10) = floor(n/10)
--     because n/2^64 * 1/4 < 1.
--
-- 32-bit case is the same, but uses 35 instead of 67.
quotRem10 :: Word -> (Word, Word)
quotRem21 n
    | finiteBitSize (0 :: Word) == 64
    = let q = (n `highMul` 14757395258967641293) `shiftR` 3 in
        (q, n - q `shiftL` 1 - q `shiftL` 3)
    | finiteBitSize (0 :: Word) == 32
    = let q = (n `highMul` 3435973837) `shiftR` 3 in
        (q, n - q `shiftL` 1 - q `shiftL` 3)
    | otherwise
    = n `quotRem` 10

highMul :: Word -> Word -> Word
highMul (W# x#) (W# y#) = W# high#
    where
        !(# high#, _ #) = timesWord2# x# y#
```

Reply:
This is not a performance critical part of the code path, it is only used on overflow, to recover the initial digits of the input stream. I did not feel it merited optimising.  Also, will this actually be significantly faster?

Reply:
In my experience it is noticeably faster, but I agree that it may be not worth it for this codepath.

Reply:
But I could multiplication by `10` by shift and add, if that's better, unless GHC already does that...  The multiplications are on the critical path when accumulating digits...